### PR TITLE
Change the error handling in the integrated auth handler

### PR DESCRIPTION
**Pull request by slewis74**
Modified the integrated auth handler to track when it issues 401 on a connection.

In the scenario where cross domain challenge doesn't result in a user, we can then detect the 2nd call and return some meaningful information to the UI.

The portal error handling had been broken for a while (since 4.0ish by the looks). This PR changes handler to push a cookie back to the browser when it redirects back to the portal, knowing that the user won't be authenticated and will get bounced to the sign in page. The sign in page has been updated to look for this cookie.

Note that the docs update for the OctoPurl is already done, see https://g.octopushq.com/TroubleshootingAD#Integrated.

Relates to OctopusDeploy/Issues#6265

### CODE REVIEW

File: source/Server/IntegratedAuthentication/IntegratedAuthenticationHost.cs [hunk 1]
```diff
@@ -68,7 +68,7 @@ public Task StartAsync()
             builder.UseHttpSys(options =>
             {
                 options.Authentication.Schemes = MapAuthenticationScheme();
-                options.Authentication.AllowAnonymous = false;
+                options.Authentication.AllowAnonymous = true;
```
Comment (johnsimons) [inline source/Server/IntegratedAuthentication/IntegratedAuthenticationHost.cs]: Is this right @slewis74 ?

File: source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs [hunk 2]
```diff
@@ -90,7 +124,29 @@ public Task HandleRequest(HttpContext context)
...
+            LoginState state = null;
+            if (context.Request.Query.TryGetValue("state", out var stateString))
+            {
+                try
+                {
+                    state = JsonConvert.DeserializeObject<LoginState>(stateString.FirstOrDefault());
+                }
+                catch (Exception e)
+                {
+                    log.Warn(e, "Invalid login state object passed to the server when setting up the NTLM challenge. Falling back to the default behaviour.");
```
Comment (michaelnoonan) [inline source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs]:
```suggestion
                    log.Warn(e, "Invalid login state object passed to the server when setting up the integrated authentication challenge. Falling back to the default behaviour.");
```

File: source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs [hunk 3]
```diff
@@ -19,43 +20,76 @@ class IntegratedAuthenticationHandler : IIntegratedAuthenticationHandler
...
             {
-                try
+                if (integratedChallengeTracker.IsConnectionKnown(context.Connection.Id))
                 {
-                    state = JsonConvert.DeserializeObject<LoginState>(stateString.FirstOrDefault());
+                    // if we've seen this connection before and the user still isn't set then something has gone
+                    // wrong with the challenge. Most likely due to cross domains now we're NETCore, https://github.com/OctopusDeploy/Issues/issues/6265
+                    var stateRedirectAfterLoginTo = state.RedirectAfterLoginTo;
+
+                    // this matches an error structure that the portal currently uses. It is not something the server it
```
Comment (michaelnoonan) [inline source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs]:
```suggestion
                    // this matches an error structure that the portal currently uses. It is not something the server
```

File: source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs [hunk 4]
```diff
@@ -19,43 +20,76 @@ class IntegratedAuthenticationHandler : IIntegratedAuthenticationHandler
...
+                    // currently knows directly about. We may make it a 1st error object at some point to help consistency.
```
Comment (michaelnoonan) [inline source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs]:
```suggestion
                    // currently knows directly about. We may make it a first-class error object at some point to help consistency.
```

File: source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs [hunk 5]
```diff
@@ -19,43 +20,76 @@ class IntegratedAuthenticationHandler : IIntegratedAuthenticationHandler
...
+                    var errorObj = new
+                    {
+                        ErrorMessage = "Authentication Error",
+                        Errors = new [] {"An error occurred with Windows authentication, possibly due to a known issue, please try using forms authentication."},
```
Comment (michaelnoonan) [inline source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs]:
```suggestion
                        Errors = new [] {"Something went wrong with Windows Integrated Authentication. Please try using forms authentication."},
```

File: source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs [hunk 6]
```diff
@@ -19,43 +20,76 @@ class IntegratedAuthenticationHandler : IIntegratedAuthenticationHandler
...
+                        DetailLinks = new[] {"https://g.octopushq.com/TroubleshootingAD#Integrated"}
+                    };
+                    
+                    // we used to use the query string to pass errors back to the portal but that was quite problematic when it came to not interfering
+                    // with deep links etc. The portal sign in page now checks for this cookie to see if an external 
+                    // auth provider is trying to return an error.
+                    context.Response.Cookies.Append("Octopus-Error", JsonConvert.SerializeObject(errorObj), new CookieOptions { MaxAge = TimeSpan.FromSeconds(15) });
```
Comment (michaelnoonan) [inline source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs]: Consider context.Request.IsHttps, state?.UsingSecureConnection

File: source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs [hunk 7]
```diff
@@ -19,43 +20,76 @@ class IntegratedAuthenticationHandler : IIntegratedAuthenticationHandler
...
+                else
                 {
-                    log.Warn(e, "Invalid login state object passed to the server when setting up the NTLM challenge. Falling back to the default behaviour.");
+                    context.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
+                    integratedChallengeTracker.SetConnectionChallengeInitiated(context.Connection.Id);
                 }
+                return Task.CompletedTask;
             }
 
+            if (integratedChallengeTracker.IsConnectionKnown(context.Connection.Id))
```
Comment (michaelnoonan) [inline source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs]: I don't think this `if` condition is required if the `SetConnectionChallengeCompleted(id)` is idempotent.

File: source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs [hunk 8]
```diff
@@ -19,43 +20,76 @@ class IntegratedAuthenticationHandler : IIntegratedAuthenticationHandler
...
-            //   1. Deep-links will not work because we don't know where the anonymous request originally wanted to go
-            //   2. Cookies may not have the Secure flag set properly when SSL Offloading is in play
-            LoginState state = null;
-            if (context.Request.Query.TryGetValue("state", out var stateString))
+            // Based on https://github.com/dotnet/runtime/blob/cf63e732fc6fb57c0ea97c1b4ca965acce46343a/src/libraries/System.Net.Security/src/System/Net/Security/NegotiateStreamPal.Windows.cs#L52
+            // we're being a bit cautious about using IsAuthenticated
+            if (string.IsNullOrWhiteSpace(context.User.Identity.Name))
             {
-                try
+                if (integratedChallengeTracker.IsConnectionKnown(context.Connection.Id))
```
Comment (michaelnoonan) [inline source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs]: Consider a positive logic flow:
1. Handle the initial challenge case first - is this a new challenge?
2. Handle the failure case second - is this a failed challenge?

Comment (michaelnoonan) [inline source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs]:
```suggestion
                if (integratedChallengeTracker.IsNewChallengeRequest(context.Connection.Id))
```

File: source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs [hunk 9]
```diff
@@ -19,43 +20,76 @@ class IntegratedAuthenticationHandler : IIntegratedAuthenticationHandler
...
+                    context.Response.Redirect(stateRedirectAfterLoginTo);
+                    
+                    // count this as complete, if the browser comes back on the same connection we'll start over 
+                    integratedChallengeTracker.SetConnectionChallengeCompleted(context.Connection.Id);
                 }
-                catch (Exception e)
+                else
                 {
-                    log.Warn(e, "Invalid login state object passed to the server when setting up the NTLM challenge. Falling back to the default behaviour.");
+                    context.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
```
Comment (michaelnoonan) [inline source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs]: Comment here: "In this case we are starting the integrated authentication challenge workflow"

File: source/Server/IntegratedAuthentication/IntegratedAuthenticationHost.cs (hunk 1 above)
Comment (michaelnoonan) [inline source/Server/IntegratedAuthentication/IntegratedAuthenticationHost.cs]: Add a comment here describing why this is important.

File: source/Server/IntegratedAuthentication/IntegratedAuthenticationHost.cs [hunk 10]
```diff
@@ -93,12 +93,6 @@ public Task StartAsync()
                          return Task.CompletedTask;
                      }
 
-                     if (string.IsNullOrWhiteSpace(context.User.Identity.Name))
```
Comment (michaelnoonan) [inline source/Server/IntegratedAuthentication/IntegratedAuthenticationHost.cs]: This has moved into the handler itself - better encapsulation and cohesion.

File: source/Server/IntegratedAuthentication/IntegratedChallengeTracker.cs [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+{
+    /// <summary>
+    /// This class is used to track integrated challenges to the browser, i.e. when we've returned a 401 to trigger a challenge.
+    /// If the same connection comes back a second time without the context's user identity being set then the challenge has failed. 
+    /// </summary>
+    class IntegratedChallengeTracker : IIntegratedChallengeTracker
+    {
+        readonly ConcurrentDictionary<string, DateTimeOffset> connections = new ConcurrentDictionary<string, DateTimeOffset>();
+        
+        public bool IsConnectionKnown(string connectionId)
```
Comment (michaelnoonan) [inline source/Server/IntegratedAuthentication/IntegratedChallengeTracker.cs]: Consider naming: "HasIntegratedAuthenticationChallengeFailed()"

File: source/Server/IntegratedAuthentication/IntegratedChallengeTracker.cs [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+                return;
+            var initiated = DateTimeOffset.Now;
+            connections.AddOrUpdate(connectionId, c => initiated, (c, i) => initiated);
+        }
+
+        public void SetConnectionChallengeCompleted(string connectionId)
+        {
+            if (!IsConnectionKnown(connectionId))
+                return;
+            connections.TryRemove(connectionId, out DateTimeOffset initiated);
```
Comment (michaelnoonan) [inline source/Server/IntegratedAuthentication/IntegratedChallengeTracker.cs]: Consider a small "TryCleanUp" - for a few lines of code, a very safe self-cleaning method feels like a good idea.

Comment (michaelnoonan) [reply source/Server/IntegratedAuthentication/IntegratedChallengeTracker.cs]: Perhaps older than 1 hour.

File: source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs (hunk 6 above)
Comment (michaelnoonan) [inline source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs]:
```suggestion
                    context.Response.Cookies.Append("Octopus-Auth-Error", JsonConvert.SerializeObject(errorObj), new CookieOptions { MaxAge = TimeSpan.FromSeconds(15) });
```

File: source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs (hunk 9 above)
Comment (michaelnoonan) [reply source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs]: Consider revoking the `Octopus-Auth-Error` cookie here.

File: source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs (hunk 7 above)
Comment (michaelnoonan) [reply source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs]: Consider revoking the `Octopus-Auth-Error` cookie here.

Comment (michaelnoonan) [reply source/Server/IntegratedAuthentication/IntegratedAuthenticationHandler.cs]: Use this code for revocation: https://github.com/OctopusDeploy/OctopusDeploy/blob/master/source/Octopus.Server/Web/Infrastructure/Authentication/AuthCookieService.cs#L284

**Pull request merged**